src/algorithms/calorimetry/CalorimeterHitDigi.cc

```diff
@@ -205,21 +213,50 @@ void CalorimeterHitDigi::process(const CalorimeterHitDigi::Input& input,
                   std::sqrt(std::pow(m_cfg.eRes[0] / std::sqrt(edep), 2) +
                             std::pow(m_cfg.eRes[1], 2) + std::pow(m_cfg.eRes[2] / (edep), 2))
             : 0;
+
     double corrMeanScale_value = corrMeanScale(leading_hit);
 
     double ped = m_cfg.pedMeanADC + m_gaussian(m_generator) * m_cfg.pedSigmaADC;
 
     // Note: both adc and tdc values must be positive numbers to avoid integer wraparound
-    unsigned long long adc =
-        std::max(std::llround(ped + edep * corrMeanScale_value * (1.0 + eResRel) /
-                                        m_cfg.dyRangeADC * m_cfg.capADC),
-                 0LL);
+    unsigned long long adc;
     unsigned long long tdc = std::llround((time + m_gaussian(m_generator) * tRes) * stepTDC);
 
-    if (edep > 1.e-3) {
+    if (readoutType == kSimpleReadout) {
+      adc = std::max(std::llround(ped + edep * corrMeanScale_value * (1.0 + eResRel) /
+                                            m_cfg.dyRangeADC * m_cfg.capADC),
+                     0LL);
+    } else if (readoutType == kPoissonPhotonReadout) {
+      const long long int n_photons_mean =
+          edep * m_cfg.lightYield * m_cfg.photonDetectionEfficiency;
+      std::poisson_distribution<> n_photons_detected_dist(n_photons_mean);
+      const long long int n_photons_detected = n_photons_detected_dist(m_generator);
+      const long long int n_max_photons =
+          m_cfg.dyRangeADC * m_cfg.lightYield * m_cfg.photonDetectionEfficiency;
+      trace("n_photons_detected {}", n_photons_detected);
+      adc = std::max(std::llround(ped + n_photons_detected * corrMeanScale_value * (1.0 + eResRel) /
+                                            n_max_photons * m_cfg.capADC),
+                     0LL);
+    } else if (readoutType == kSipmReadout) {
+      const long long int n_photons = edep * m_cfg.lightYield;
```

Comment:
I guess it depends on what resolution smearing be eResRel represents with this change... But shouldn't SiPM saturation include resolution smearing by (1.0 + eResRel)? If so
  n_photons_mean = edep * (1.0 + eResRel) * m_cfg.lightYield * m_cfg.photonDetectionEfficiency;
then
  adc = std::max(std::llround(ped + n_photons_detected * corrMeanScale_value / n_max_photons * m_cfg.capADC), 0LL);

The same for PoissonPhotonReadout too.

Reply:
So my understanding is that eResRel is used to fudge calorimeter resolutions to expected ones. Then it makes sense to apply it after all non-linearities so that those parameters can be extracted by quadrature substraction of observed resolution from a target resolution. Does that make sense? I'm happy to reorder it the other way, if you prefer. No other detector uses eRelRel (*except B0 which does photon smearing with it, so it won't enable both*).